File: src/lib/data/query.ts
import type { PackFilter } from '../schemas/pack.ts';
import type { Question, VariantName } from '../schemas/question.ts';
import type { LoadedDataset } from './load.ts';

/**
 * Query the question pool using a PackFilter.
 * All filter operators are ANDed together.
 * Returns question IDs matching all criteria.
 */
export function queryPool(ds: LoadedDataset, filter: PackFilter): string[] {
	const { kinds, tags_all, tags_any, tags_none, variants_any, limit } = filter;

	const questions = [...ds.questions.values()];

	const matched = questions.filter((entry) => {
		const q = entry.item;

		if (kinds && kinds.length > 0 && !kinds.includes(q.kind)) return false;

		if (tags_all && tags_all.length > 0) {
			if (!tags_all.every((tag) => q.tags.includes(tag))) return false;
		}

		if (tags_any && tags_any.length > 0) {
			if (!tags_any.some((tag) => q.tags.includes(tag))) return false;
		}

		if (tags_none && tags_none.length > 0) {
			if (tags_none.some((tag) => q.tags.includes(tag))) return false;
		}

		if (variants_any && variants_any.length > 0) {
			const defined = getVariantNames(q);
			if (!variants_any.some((vname) => defined.has(vname))) return false;
		}

		return true;
	});

	const ids = matched.map((entry) => entry.item.id);
	return limit ? ids.slice(0, limit) : ids;
}

function getVariantNames(q: Question): Set<VariantName> {
	if (q.kind === 'order') return new Set();
	const variants = q.content.variants as Record<string, unknown>;
	return new Set(Object.keys(variants).filter((k) => variants[k] !== undefined) as VariantName[]);
}

File: api/src/data/query.rs
//! Pool query engine — filter the question pool using a `PackFilter`.
//!
//! All filter operators are ANDed together. Mirrors `query.ts`.

use super::types::*;

/// Returns question IDs matching all filter criteria.
pub fn query_pool(ds: &LoadedDataset, filter: &PackFilter) -> Vec<String> {
    let matched: Vec<&Entry<Question>> = ds
        .questions
        .values()
        .filter(|entry| matches_filter(&entry.item, filter))
        .collect();

    let ids: Vec<String> = matched.iter().map(|e| e.item.id().to_owned()).collect();

    match filter.limit {
        Some(limit) => ids.into_iter().take(limit).collect(),
        None => ids,
    }
}

fn matches_filter(q: &Question, f: &PackFilter) -> bool {
    if let Some(ref kinds) = f.kinds {
        if !kinds.is_empty() && !kinds.contains(&q.kind()) {
            return false;
        }
    }

    let tags = q.tags();

    if let Some(ref tags_all) = f.tags_all {
        if !tags_all.is_empty() && !tags_all.iter().all(|t| tags.contains(t)) {
            return false;
        }
    }

    if let Some(ref tags_any) = f.tags_any {
        if !tags_any.is_empty() && !tags_any.iter().any(|t| tags.contains(t)) {
            return false;
        }
    }

    if let Some(ref tags_none) = f.tags_none {
        if tags_none.iter().any(|t| tags.contains(t)) {
            return false;
        }
    }

    if let Some(ref variants_any) = f.variants_any {
        if !variants_any.is_empty() {
            let defined = q.variant_names();
            if !variants_any.iter().any(|v| defined.contains(v)) {
                return false;
            }
        }
    }

    true
}
